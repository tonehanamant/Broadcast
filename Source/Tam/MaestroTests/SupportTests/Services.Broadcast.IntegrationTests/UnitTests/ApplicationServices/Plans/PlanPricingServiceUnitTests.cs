using Common.Services;
using Common.Services.Repositories;
using Hangfire;
using Moq;
using NUnit.Framework;
using Services.Broadcast.ApplicationServices;
using Services.Broadcast.BusinessEngines;
using Services.Broadcast.Cache;
using Services.Broadcast.Clients;
using Services.Broadcast.Entities.Enums;
using Services.Broadcast.Entities.Plan.Pricing;
using Services.Broadcast.Repositories;
using System;

namespace Services.Broadcast.IntegrationTests.UnitTests.ApplicationServices.Plans
{
    [TestFixture]
    public class PlanPricingServiceUnitTests
    {
        private readonly Mock<IDataRepositoryFactory> _DataRepositoryFactoryMock;
        private readonly Mock<ISpotLengthEngine> _SpotLengthEngineMock;
        private readonly Mock<IPricingApiClient> _PricingApiClientMock;
        private readonly Mock<IBackgroundJobClient> _BackgroundJobClientMock;
        private readonly Mock<IPlanPricingInventoryEngine> _PlanPricingInventoryEngineMock;
        private readonly Mock<IBroadcastLockingManagerApplicationService> _BroadcastLockingManagerApplicationServiceMock;

        public PlanPricingServiceUnitTests()
        {
            _DataRepositoryFactoryMock = new Mock<IDataRepositoryFactory>();
            _SpotLengthEngineMock = new Mock<ISpotLengthEngine>();
            _PricingApiClientMock = new Mock<IPricingApiClient>();
            _BackgroundJobClientMock = new Mock<IBackgroundJobClient>();
            _PlanPricingInventoryEngineMock = new Mock<IPlanPricingInventoryEngine>();
            _BroadcastLockingManagerApplicationServiceMock = new Mock<IBroadcastLockingManagerApplicationService>();
        }

        [Test]
        [TestCase(BackgroundJobProcessingStatus.Queued)]
        [TestCase(BackgroundJobProcessingStatus.Processing)]
        public void CanNotQueuePricingJobWhenThereIsOneActive(BackgroundJobProcessingStatus status)
        {
            const string expectedMessage = "The pricing model is already running for the plan";

            var planRepositoryMock = new Mock<IPlanRepository>();
            planRepositoryMock
                .Setup(x => x.GetLatestPricingJob(It.IsAny<int>()))
                .Returns(new PlanPricingJob { Status = status });

            _DataRepositoryFactoryMock
                .Setup(x => x.GetDataRepository<IPlanRepository>())
                .Returns(planRepositoryMock.Object);

            _BroadcastLockingManagerApplicationServiceMock
                .Setup(x => x.GetNotUserBasedLockObjectForKey(It.IsAny<string>()))
                .Returns(new object());

            var service = new PlanPricingService(
                _DataRepositoryFactoryMock.Object,
                _SpotLengthEngineMock.Object,
                _PricingApiClientMock.Object,
                _BackgroundJobClientMock.Object,
                _PlanPricingInventoryEngineMock.Object,
                _BroadcastLockingManagerApplicationServiceMock.Object);

            var exception = Assert.Throws<Exception>(() => service.QueuePricingJob(
                planPricingParametersDto: new PlanPricingParametersDto(), 
                currentDate: new DateTime(2019, 10, 23)));

            Assert.AreEqual(expectedMessage, exception.Message);
        }
    }
}
