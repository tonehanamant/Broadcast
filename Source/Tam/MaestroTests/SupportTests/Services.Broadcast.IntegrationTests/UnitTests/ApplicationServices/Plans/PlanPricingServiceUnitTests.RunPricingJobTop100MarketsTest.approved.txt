[
  {
    "contract": [
      {
        "cpm_goal": 80.00,
        "daypart_weighting": [
          {
            "daypart_goal": 0.6,
            "daypart_id": 1
          },
          {
            "daypart_goal": 0.4,
            "daypart_id": 2
          }
        ],
        "frequency_cap_spots": 10,
        "frequency_cap_time": 1.0,
        "frequency_cap_unit": "day",
        "impression_goal": 150.0,
        "market_coverage_goal": 0.8,
        "share_of_voice": [
          {
            "market_code": 403,
            "market_goal": 0.0004743
          },
          {
            "market_code": 202,
            "market_goal": 0.0002942
          },
          {
            "market_code": 102,
            "market_goal": 0.13
          }
        ],
        "week_id": 100
      },
      {
        "cpm_goal": 48.000,
        "daypart_weighting": [
          {
            "daypart_goal": 0.6,
            "daypart_id": 1
          },
          {
            "daypart_goal": 0.4,
            "daypart_id": 2
          }
        ],
        "frequency_cap_spots": 10,
        "frequency_cap_time": 1.0,
        "frequency_cap_unit": "day",
        "impression_goal": 250.0,
        "market_coverage_goal": 0.8,
        "share_of_voice": [
          {
            "market_code": 403,
            "market_goal": 0.0004743
          },
          {
            "market_code": 202,
            "market_goal": 0.0002942
          },
          {
            "market_code": 102,
            "market_goal": 0.13
          }
        ],
        "week_id": 101
      },
      {
        "cpm_goal": 120.00,
        "daypart_weighting": [
          {
            "daypart_goal": 0.6,
            "daypart_id": 1
          },
          {
            "daypart_goal": 0.4,
            "daypart_id": 2
          }
        ],
        "frequency_cap_spots": 10,
        "frequency_cap_time": 1.0,
        "frequency_cap_unit": "day",
        "impression_goal": 100.0,
        "market_coverage_goal": 0.8,
        "share_of_voice": [
          {
            "market_code": 403,
            "market_goal": 0.0004743
          },
          {
            "market_code": 202,
            "market_goal": 0.0002942
          },
          {
            "market_code": 102,
            "market_goal": 0.13
          }
        ],
        "week_id": 102
      }
    ],
    "inventory": [
      {
        "cost": 50.0,
        "daypart_id": 15,
        "id": 1,
        "impressions": 1000.0,
        "market_code": 101,
        "percentage_of_us": 0.00101,
        "spot_days": 1,
        "spot_hours": 1.0,
        "station_id": 5,
        "week_id": 100
      },
      {
        "cost": 50.0,
        "daypart_id": 15,
        "id": 1,
        "impressions": 1000.0,
        "market_code": 101,
        "percentage_of_us": 0.00101,
        "spot_days": 1,
        "spot_hours": 1.0,
        "station_id": 5,
        "week_id": 101
      },
      {
        "cost": 50.0,
        "daypart_id": 15,
        "id": 1,
        "impressions": 1000.0,
        "market_code": 101,
        "percentage_of_us": 0.00101,
        "spot_days": 1,
        "spot_hours": 1.0,
        "station_id": 5,
        "week_id": 102
      },
      {
        "cost": 60.0,
        "daypart_id": 16,
        "id": 2,
        "impressions": 1500.0,
        "market_code": 100,
        "percentage_of_us": 0.001,
        "spot_days": 2,
        "spot_hours": 2.0,
        "station_id": 6,
        "week_id": 100
      }
    ]
  }
]